File: lib/python/openlinktoken-core-ai/src/test/openlinktoken/core/ai/tokens/ml1_onnx_signature_provider_test.py
import pytest

from openlinktoken.core.ai.tokens.ml1_inference_config import ML1InferenceConfig
from openlinktoken.core.ai.tokens.ml1_onnx_signature_provider import (
    ML1OnnxSignatureProvider,
    _build_rotation_signature,
    _compute_blocking_key,
    _hash_rotation_values,
)
from openlinktoken.core.ai.tokens.ml1_token import ML1Token
from openlinktoken.core.ai.tokens.rotation_config import RotationConfig
from openlinktoken.tokens.token import Token
from openlinktoken.tokens.token_generator_result import TokenGeneratorResult


@pytest.fixture(autouse=True)
def reset_runtime_config():
    """Restore process-wide ML1 and rotation settings after each test."""

    def restore():
        ML1InferenceConfig.configure(
            True,
            ML1InferenceConfig.DEFAULT_MODEL_PATH,
            ML1InferenceConfig.DEFAULT_TOKENIZER_PATH,
            ML1InferenceConfig.DEFAULT_MAX_SEQUENCE_LENGTH,
            ML1InferenceConfig.DEFAULT_BATCH_SIZE,
            ML1InferenceConfig.DEFAULT_NUM_THREADS,
        )
        RotationConfig.configure(enable=True, rotation_iv=RotationConfig.DEFAULT_IV)

    restore()
    yield
    restore()


def test_rotation_hash_uses_hashed_t1_blocking_key():
    """Verify rotation values are hashed with the normalized T1 blocking key."""
    person_attributes = {
        "BirthDate": "1989-05-25",
        "FirstName": "Chelsea",
        "LastName": "Meister",
        "PostalCode": "06582",
        "Sex": "Female",
    }

    blocking_key = _compute_blocking_key(person_attributes)

    assert blocking_key == "f016a96ba8552da8c9d7ac327f91081e22740f0ddd71dc372fa4dbba2ca34253"
    assert _hash_rotation_values(["99 100 100 101"], blocking_key) == [
        "4ff691600f8c2df6142c405cbcd6f166a588ba83bd93ba6f028e082ef99decd8"
    ]


def test_rotation_hash_without_blocking_key_returns_no_token():
    """Rotation values must not be emitted when no T1 blocking key is available."""
    assert _hash_rotation_values(["99 100 100 101"], None) is None


def test_rotation_signature_without_blocking_key_returns_blank_token():
    """A missing T1 blocking key should produce the canonical blank token."""
    assert _build_rotation_signature(["99 100 100 101"], None) == Token.BLANK


def test_provider_uses_ml1_token_identifier():
    """Provider and token definition must share the ML1 registry identifier."""
    assert ML1OnnxSignatureProvider().get_token_id() == ML1Token.TOKEN_ID


def test_build_ml1_payload_normalizes_fields_in_definition_order():
    """ML1 payloads should normalize values and preserve the model field order."""
    person_attributes = {
        "PostalCode": "95123",
        "BirthDate": "1990-07-09",
        "FirstName": " Alice ",
        "LastName": " Smith ",
        "Sex": "female",
    }
    result = TokenGeneratorResult()

    payload = ML1OnnxSignatureProvider().build_ml1_payload(person_attributes, result)

    assert payload == (
        '{"PostalCode": "95123", "Birthdate": "1990-07-09", "GivenName": "Alice", '
        '"Surname": "Smith", "Gender": "Female"}'
    )
    assert result.invalid_attributes == set()


def test_build_ml1_payload_records_missing_required_field():
    """Missing required ML1 fields should produce no payload and record the field."""
    person_attributes = {
        "PostalCode": "95123",
        "BirthDate": "1990-07-09",
        "FirstName": "Alice",
        "LastName": "Smith",
    }
    result = TokenGeneratorResult()

    payload = ML1OnnxSignatureProvider().build_ml1_payload(person_attributes, result)

    assert payload is None
    assert result.invalid_attributes == {"Sex"}


def test_build_ml1_payload_records_invalid_field():
    """Invalid required ML1 fields should produce no payload and record the field."""
    person_attributes = {
        "PostalCode": "95123",
        "BirthDate": "1990-07-09",
        "FirstName": "Alice",
        "LastName": "Smith",
        "Sex": "unknown",
    }
    result = TokenGeneratorResult()

    payload = ML1OnnxSignatureProvider().build_ml1_payload(person_attributes, result)

    assert payload is None
    assert result.invalid_attributes == {"Sex"}


@pytest.mark.parametrize("enabled", [True, False])
def test_provider_enabled_state_matches_configuration(enabled):
    """Provider enablement should reflect the process-wide ML1 configuration."""
    ML1InferenceConfig.configure(
        enabled,
        ML1InferenceConfig.DEFAULT_MODEL_PATH,
        ML1InferenceConfig.DEFAULT_TOKENIZER_PATH,
        ML1InferenceConfig.DEFAULT_MAX_SEQUENCE_LENGTH,
        ML1InferenceConfig.DEFAULT_BATCH_SIZE,
        ML1InferenceConfig.DEFAULT_NUM_THREADS,
    )

    assert ML1OnnxSignatureProvider().is_enabled() is enabled


def test_generate_batch_returns_empty_signatures_for_all_invalid_rows():
    """A batch with no valid payloads should not initialize or call ONNX."""
    result = ML1OnnxSignatureProvider().generate_batch([{}, {}])

    assert result.signatures == [None, None]


def test_ml1_token_definition_is_empty():
    """ML1 input is the canonical payload rather than attribute expressions."""
    assert ML1Token().get_definition() == []

File: lib/java/openlinktoken-core-ai/src/test/java/org/openlinktoken/core/ai/tokens/ML1OnnxSignatureProviderTest.java
/* SPDX-License-Identifier: MIT */
package org.openlinktoken.core.ai.tokens;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.openlinktoken.core.ai.tokentransformer.rotation.RotationEmbeddingTransformer;
import org.openlinktoken.core.ai.tokens.definitions.ML1Token;
import org.openlinktoken.tokens.Token;
import org.openlinktoken.tokens.TokenGeneratorResult;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Tests ML1 payload construction, signatures, rotation hashing, and batch behavior.
 */
class ML1OnnxSignatureProviderTest {

    private ML1OnnxSignatureProvider provider;

    @Test
    void getTokenId_matchesMl1TokenIdentifier() {
        assertEquals(ML1Token.TOKEN_ID, provider.getTokenId());
    }

    @BeforeEach
    void setUp() {
        provider = new ML1OnnxSignatureProvider();
    }

    @AfterEach
    void tearDown() throws Exception {
        RotationConfig.configure(
                true,
                RotationConfig.DEFAULT_IV,
                RotationConfig.DEFAULT_ROTATION_COUNT,
                RotationConfig.DEFAULT_HASH_DIMENSION,
                RotationConfig.DEFAULT_BIN_WIDTH,
                RotationConfig.DEFAULT_MIN_VAL,
                RotationConfig.DEFAULT_MAX_VAL);
        ML1InferenceConfig.configure(
                true,
                ML1InferenceConfig.DEFAULT_MODEL_PATH,
                ML1InferenceConfig.DEFAULT_TOKENIZER_PATH,
                ML1InferenceConfig.DEFAULT_MAX_SEQUENCE_LENGTH,
                ML1InferenceConfig.DEFAULT_BATCH_SIZE,
                ML1InferenceConfig.DEFAULT_NUM_THREADS);
        resetRotationTransformer();
    }

    // -----------------------------------------------------------------------
    // ML1 payload construction and provider state
    // -----------------------------------------------------------------------

    @Test
    void buildMl1Payload_validAttributes_preservesFieldOrderAndNormalization() {
        Map<String, String> attrs = new HashMap<>();
        attrs.put("PostalCode", "95123");
        attrs.put("BirthDate", "1990-07-09");
        attrs.put("FirstName", " Alice ");
        attrs.put("LastName", " Smith ");
        attrs.put("Sex", "female");

        String payload = provider.buildMl1Payload(attrs, new TokenGeneratorResult());

        assertEquals(
                "{\"PostalCode\": \"95123\", \"Birthdate\": \"1990-07-09\", "
                        + "\"GivenName\": \"Alice\", \"Surname\": \"Smith\", \"Gender\": \"Female\"}",
                payload);
    }

    @Test
    void buildMl1Payload_missingRequiredField_returnsNull() {
        Map<String, String> attrs = new HashMap<>();
        attrs.put("PostalCode", "95123");
        attrs.put("BirthDate", "1990-07-09");
        attrs.put("FirstName", "Alice");
        attrs.put("LastName", "Smith");

        assertNull(provider.buildMl1Payload(attrs, new TokenGeneratorResult()));
    }

    @Test
    void buildMl1Payload_invalidRequiredField_recordsInvalidAttribute() {
        Map<String, String> attrs = new HashMap<>();
        attrs.put("PostalCode", "95123");
        attrs.put("BirthDate", "1990-07-09");
        attrs.put("FirstName", "Alice");
        attrs.put("LastName", "Smith");
        attrs.put("Sex", "unknown");
        TokenGeneratorResult result = new TokenGeneratorResult();

        assertNull(provider.buildMl1Payload(attrs, result));
        assertTrue(result.getInvalidAttributes().contains("Sex"), result.getInvalidAttributes().toString());
    }

    @Test
    void isEnabled_reflectsInferenceConfiguration() {
        ML1InferenceConfig.configure(false, "", "", 128, 64, 1);
        assertTrue(!provider.isEnabled());

        ML1InferenceConfig.configure(true, "", "", 128, 64, 1);
        assertTrue(provider.isEnabled());
    }

    @Test
    void generateBatch_allInvalidRows_returnsNullForEachRow() {
        List<Map<String, String>> rows = List.of(Map.of(), Map.of());

        List<String> signatures = provider.generateBatch(rows).signatures();

        assertEquals(Arrays.asList(null, null), signatures);
    }

    // -----------------------------------------------------------------------
    // computeT1Signature
    // -----------------------------------------------------------------------

    @Test
    void computeT1Signature_validAttributes_returnsExpectedSignature() {
        Map<String, String> attrs = new HashMap<>();
        attrs.put("LastName", "Wright");
        attrs.put("FirstName", "Robert");
        // SexAttribute validates against ^([Mm](ale)?|[Ff](emale)?)$ — use mixed-case input
        attrs.put("Sex", "Female");
        attrs.put("BirthDate", "1990-07-09");

        String sig = provider.computeT1Signature(attrs);

        // normalize("Female") → "Female"; T|U → "FEMALE"
        assertEquals("WRIGHT|R|FEMALE|1990-07-09", sig);
    }

    @Test
    void computeT1Signature_nullMap_returnsNull() {
        assertNull(provider.computeT1Signature(null));
    }

    @Test
    void computeT1Signature_missingLastName_returnsNull() {
        Map<String, String> attrs = new HashMap<>();
        attrs.put("FirstName", "Robert");
        attrs.put("Sex", "FEMALE");
        attrs.put("BirthDate", "1990-07-09");

        assertNull(provider.computeT1Signature(attrs));
    }

    @Test
    void computeT1Signature_missingFirstName_returnsNull() {
        Map<String, String> attrs = new HashMap<>();
        attrs.put("LastName", "Wright");
        attrs.put("Sex", "FEMALE");
        attrs.put("BirthDate", "1990-07-09");

        assertNull(provider.computeT1Signature(attrs));
    }

    @Test
    void computeT1Signature_missingSex_returnsNull() {
        Map<String, String> attrs = new HashMap<>();
        attrs.put("LastName", "Wright");
        attrs.put("FirstName", "Robert");
        attrs.put("BirthDate", "1990-07-09");

        assertNull(provider.computeT1Signature(attrs));
    }

    @Test
    void computeT1Signature_missingBirthDate_returnsNull() {
        Map<String, String> attrs = new HashMap<>();
        attrs.put("LastName", "Wright");
        attrs.put("FirstName", "Robert");
        attrs.put("Sex", "Female");

        assertNull(provider.computeT1Signature(attrs));
    }

    @Test
    void computeT1Signature_invalidBirthDate_returnsNull() {
        Map<String, String> attrs = new HashMap<>();
        attrs.put("LastName", "Wright");
        attrs.put("FirstName", "Robert");
        attrs.put("Sex", "Female");
        attrs.put("BirthDate", "not-a-date");

        assertNull(provider.computeT1Signature(attrs));
    }

    @Test
    void computeT1Signature_lowercaseInputsAreNormalized() {
        Map<String, String> attrs = new HashMap<>();
        attrs.put("LastName", "  smith  ");
        attrs.put("FirstName", "  alice  ");
        attrs.put("Sex", "male");
        attrs.put("BirthDate", "2000-01-15");

        String sig = provider.computeT1Signature(attrs);

        assertEquals("SMITH|A|MALE|2000-01-15", sig);
    }

    @Test
    void rotationConfig_defaultIv_matchesPythonParityValue() {
        assertEquals("openlinktoken-ml1-v1", RotationConfig.DEFAULT_IV);
    }

    @Test
    void getOrCreateTransformer_usesConfiguredRotationParameters() throws Exception {
        RotationConfig.configure(true, "", 3, 2, 0.25, -2.5, 2.5, new double[] { 1.5, -0.5, 0.0, 2.0 });
        resetRotationTransformer();

        RotationEmbeddingTransformer transformer = getRotationTransformer(4);

        assertEquals(RotationConfig.DEFAULT_IV, readField(transformer, "iv"));
        assertEquals(3, readField(transformer, "rotationCount"));
        assertEquals(2, readField(transformer, "hashDimension"));
        assertEquals(0.25d, (double) readField(transformer, "binWidth"));
        assertEquals(-2.5d, (double) readField(transformer, "minVal"));
        assertEquals(2.5d, (double) readField(transformer, "maxVal"));
        double[] bias = (double[]) readField(transformer, "bias");
        assertEquals(List.of(1.5, -0.5, 0.0, 2.0), List.of(bias[0], bias[1], bias[2], bias[3]));
    }

    // -----------------------------------------------------------------------
    // hashRotationValues
    // -----------------------------------------------------------------------

    @Test
    void hashRotationValues_returnsOneHexDigestPerInput() {
        List<String> rotationValues = List.of("94 104 96 97", "12 34 56 78");
        List<String> result = provider.hashRotationValues(rotationValues, "WRIGHT|R|FEMALE|1990-07-09");

        assertEquals(2, result.size());
        for (String digest : result) {
            // SHA-256 produces 32 bytes → 64 hex chars
            assertEquals(64, digest.length(), "Expected 64-char hex digest");
            assertTrue(digest.matches("[0-9a-f]+"), "Digest must be lowercase hex");
        }
    }

    @Test
    void hashRotationValues_deterministicOutput() {
        List<String> rotationValues = List.of("94 104 96 97");
        String key = "WRIGHT|R|FEMALE|1990-07-09";

        List<String> first = provider.hashRotationValues(rotationValues, key);
        List<String> second = provider.hashRotationValues(rotationValues, key);

        assertEquals(first, second, "Hash must be deterministic");
    }

    @Test
    void hashRotationValues_differentKeysProduceDifferentDigests() {
        List<String> rotationValues = List.of("94 104 96 97");

        String digest1 = provider.hashRotationValues(rotationValues, "WRIGHT|R|FEMALE|1990-07-09").get(0);
        String digest2 = provider.hashRotationValues(rotationValues, "SMITH|A|MALE|2000-01-15").get(0);

        assertTrue(!digest1.equals(digest2), "Different keys must produce different digests");
    }

    @Test
    void hashRotationValues_knownVector() {
        String rawT1 = "MEISTER|C|FEMALE|1989-05-25";
        String blockingKey = provider.computeT1BlockingKey(rawT1);

        assertEquals("f016a96ba8552da8c9d7ac327f91081e22740f0ddd71dc372fa4dbba2ca34253", blockingKey);
        assertEquals(
                List.of("4ff691600f8c2df6142c405cbcd6f166a588ba83bd93ba6f028e082ef99decd8"),
                provider.hashRotationValues(List.of("99 100 100 101"), blockingKey));
    }

    @Test
    void hashRotationValues_missingBlockingKey_returnsNull() {
        assertNull(provider.hashRotationValues(List.of("99 100 100 101"), null));
    }

    @Test
    void buildRotationSignature_missingBlockingKey_returnsBlankToken() {
        assertEquals(
                Token.BLANK,
                provider.buildRotationSignature(List.of("99 100 100 101"), null));
    }

    private static RotationEmbeddingTransformer getRotationTransformer(int embeddingDim) throws Exception {
        Method getOrCreateTransformer = ML1OnnxSignatureProvider.class.getDeclaredMethod(
                "getOrCreateTransformer",
                int.class);
        getOrCreateTransformer.setAccessible(true);
        return (RotationEmbeddingTransformer) getOrCreateTransformer.invoke(null, embeddingDim);
    }

    private static Object readField(Object target, String fieldName) throws Exception {
        Field field = target.getClass().getDeclaredField(fieldName);
        field.setAccessible(true);
        return field.get(target);
    }

    private static void resetRotationTransformer() throws Exception {
        Field transformerField = ML1OnnxSignatureProvider.class.getDeclaredField("rotationTransformer");
        transformerField.setAccessible(true);
        transformerField.set(null, null);
    }
}
